protocol/contracts/templegold/TempleGoldVesting.sol

```diff
@@ -0,0 +1,339 @@
+     */
+    function setPaymentToken(address /*_token*/) public pure override {
+        revert NotImplemented();
+    }
+
+    /**
+     * @dev Computes the next vesting schedule identifier for a given account address.
+     */
+    function computeNextVestingScheduleIdForHolder(
+        address _recipient
+    ) public view override returns (bytes32) {
+        return
+            computeVestingScheduleIdForAddressAndIndex(
+                _recipient,
+                holdersVestingCount[_recipient]
+            );
+    }
+
+    /**
+     * @notice Get vesting Id at index from list of vesting Ids
+     * @param _index Index
+     * @return id Bytes32 Id
+     */
+    function getVestingIdAtIndex(
+        uint256 _index
+    ) external view override returns (bytes32 id) {
+        /// @dev revert with an array out-of-bounds access if only `id = _vestingIds.at(_index);` is used
+        if (_vestingIds.length() == 0) { return id; }
+        id = _vestingIds.at(_index);
+    }
+
+    /**
+     * @notice Release tokens vested at current block timestamp. Caller must be recipient of vest
+     * @param _vestingId Vesting Id
+     */
+    function release(bytes32 _vestingId) external override {
+        if(!vestingIdExists(_vestingId)) { revert CommonEventsAndErrors.InvalidParam(); }
+        VestingSchedule storage _schedule = schedules[_vestingId];
+        uint256 _releasableAmount  = _validateRelease(_schedule);
+        _release(_schedule, _releasableAmount);
+        emit Released(_vestingId, msg.sender, _releasableAmount);
+    }
+
+    /**
+     * @dev Computes the vesting schedule identifier for an address and an index.
+     */
+    function computeVestingScheduleIdForAddressAndIndex(
+        address holder,
+        uint256 index
+    ) public pure override returns (bytes32) {
+        return keccak256(abi.encodePacked(holder, index));
+    }
+
+    /**
+     * @notice Check if vesting Id exists
+     * @param _id Vesting Id
+     * @return Bool
+     */
+     function vestingIdExists(bytes32 _id) public view returns (bool) {
+        return _vestingIds.contains(_id);
+    }
+
+    /**
+     * @notice Check if vesting is revoked
+     * @param _vestingId Vesting Id
+     * @return Bool
+     */
+    function isVestingRevoked(bytes32 _vestingId) external view override returns (bool) {
+        if(!vestingIdExists(_vestingId)) { return true; }
+        return schedules[_vestingId].revoked;
+    }
+
+    /**
+     * @notice Get vesting schedule
+     * @param _vestingId Vesting Id
+     * @return schedule VestingSchedule 
+     */
+    function getSchedule(bytes32 _vestingId) external view override returns (VestingSchedule memory schedule) {
+        schedule = schedules[_vestingId];
+    }
+
+    /**
+     * @notice Returns the last vesting schedule for a given holder address.
+     * @param _recipient Recipient address
+     * @return VestingSchedule
+     */
+    function getLastVestingScheduleForHolder(
```

Comment:
I don't see tests for these...need to make sure you have tests for **everything**, and that you cover all edge cases please.